### DISCUSSION TITLE
A condition in If CodeQL formulas that requires comparison of different types

### DISCUSSION THREAD

**Discussion by Sim4n6** [github/codeql#10618] [Q&A] [answered]
I have this verification to be done using the CodeQL formulas `If`.
I need to check whether `arg` is equal to `None` or `[]` (empty list) or `C` (random stuff).

```codeql

from DataFlow::Node sink, DataFlow::Node arg
where
    if 
          arg instanceof None
          or
          arg instanceof ListObject and arg.(ListObject).getLength() = 0
          or
          arg = C
    then sink = A
    else sink = B
  )
select sink, "sink" + sink.getLocation()
``` 
I wonder how to guess a common CodeQL type that would fit `arg`? how to fix this, please?  

Sorry guys if I ask too many questions in here, but I love CodeQL and I'm a bit lost...

**Comment by Sim4n6** [comment]
To be more precise `arg` can take values as such:
-  `None`
- `[]`
- `[Obj1,Obj1,...]`
- `f(Obj1)`

any help is a welcome ?

**Comment by yoff** [comment, marked as answer]
Could you describe what you are trying to achieve overall? It looks like you might want one data flow query for the case `sink = A` with sources being `None` and the empty list, and another for the case `sink = B`. But I am not sure when the case `sink = B` is supposed to kick in, since the first case also includes `C` (random stuff).

If it is all purely syntactic, you could use `AstNode` as the common type...

**Reply by Sim4n6** [reply]
I figure it out 👍🏾 
```codeql 
if
        arg.asCfgNode() instanceof NameConstantNode or
        arg.asCfgNode() instanceof ListNode
      then sink = A
      else sink = B
```

Now how to check `arg` is an empty list ... please?

**Reply by tausbn** [reply]
To check whether a `ListNode` `l` corresponds to an empty list, you can use `not exists(l.getAnElement())`.

**Reply by Sim4n6** [reply]
very simple and elegant, thank you 👍🏾
